### DISCUSSION TITLE
3D Binary classification using monai, pytorch based framework

### DISCUSSION THREAD

**Discussion by Harris91-lee** [Project-MONAI/MONAI#6016] [General]
Hello everyone, 
My name is Harris. In the past few months i started using Monai framework and i used SEGRESNET model to do single and Multi label segementation. It worked just fine. 

However, now im trying to do Binary classification using monai framework as well. I have volume data of good cases in one folder and bad cases in another folder. I also have a txt file that contains the file names and their labels, for ex:
***.nii.gz    0 
***.nii.gz    0
***.nii.gz    0
***.nii.gz    1
***.nii.gz    1
...... so on 

So my question in short how can i prepare my data and their binary labels? 
let's say the volume data will be [1,1,128,128,128], where: 1 referese to number of volume, the other 1 is the channel number, and the 128 is volume dimensions. Should i add the label at the end? so it will be like [1,1,128,128,128,1]? or should i slice the volume to 2D first then do data preparation and model training? 
i found same examples for cnn,vgg16,densenet121, ... but mostly they are 2D. So im posting here my question. 

I will be grateful to hear some suggestions from you. 

Thank you.

**Comment by KumoLiu** [comment]
Hi @Harris91-lee, I think you can find some 3d classification tutorials in [this fold](https://github.com/Project-MONAI/tutorials/tree/main/3d_classification). You can use a dictionary-based transform or array-based transform, it's up to you. The difference is that you should use `ImageDataset` for the array-based and `monai.data.Dataset` for the dict-based. You could find more details in the link I posted.
https://github.com/Project-MONAI/tutorials/blob/4d4c565c681fd7df0b01889c9d599b9b73e0b401/3d_classification/torch/densenet_training_dict.py#L33-L60

https://github.com/Project-MONAI/tutorials/blob/4d4c565c681fd7df0b01889c9d599b9b73e0b401/3d_classification/torch/densenet_training_array.py#L32-L57

Hope it can help you, if faced any issues, feel free to ask me, thanks!

**Reply by Harris91-lee** [reply]
Thank you alot the implementation look nice :) @KumoLiu 
I will try it out and come back to you. So trying either dictionary-based transform or array-based transform is prefrence matter as i understood.

**Reply by KumoLiu** [reply]
I would like to say both is ok~

**Reply by Harris91-lee** [reply]
So while training it does only save model with epoch 1 as the best model, and when inferencing it doesnt have always same answer (same class). 
i saved other models like 50th and 99th epoch and it says it has acc =1.0 and average loss 0.0. i tested all models i got so far and the results aren't good. Seems like biased results? Any suggestions? I'm using Ultra-sound volumes 128x128x128 and in the transforms there is a resizing process to 96 but i dont think it's the issue? 

i'm try to do Binary classifiaction whether its first trimester or second/third trimester. 

many thanks in advance @KumoLiu

**Reply by KumoLiu** [reply]
Hi @Harris91-lee, if you save epoch 1 as the best model, I would like to say maybe your model didn't converge at all. Several things you could do here:

- If you can't ensure that you doing the transforms in the right way, examine the data before sending it to the network to make sure it makes sense. Since I'm not familiar with the Ultra-sound data, I'm not sure if the scale intensity operation in the tutorial is appropriate, but you can plot the data and examine it before entering the network.
- If the data is right, try to tune the hyperparameters, such as learning rate, batch size, optimizer, and so on.
- Also I noticed that you said biased issue, if the proportion of positive and negative samples in your data is very unbalanced, you should try to sample as evenly as possible, or apply a larger penalty for large sample sizes when calculating the loss.

Hope it can help you, thanks!

**Reply by Harris91-lee** [reply]
Thank you @KumoLiu for your fast response. I appreciate that.
Regrading: 
- point1, i should check that right. But now you mean the scaleIntensityd is appropriate for MR images? 
- point2, I'm kinda doing that actually now and im noticing some imporvement when the dataset is smaller. like using 1500 volume for both classes vs using 100 volume for both classes using same hyper-paramter got me some improvement with the smaller set. 
- point3, i think i use even dataset. i meant to write "overfitted" my bad. 

Thank you again.

**Reply by KumoLiu** [reply]
For MR data, performing z-score normalization for each patient's data individually may be appropriate referring to https://www.nature.com/articles/s41592-020-01008-z.

**Reply by Harris91-lee** [reply]
Hello @KumoLiu :)
So im getting good results thanks to you :) 
The accuracy is between 92% to 96% depending on the testset. 
I wonder if i can improve it to 98% or 99%. Like what will be a good approach to do that? 
i already tried batch-size, epoch, resizing 64,96,128, do you have another suggestion? Dataset size interestingly shouldn't big big, i mean now im using dataset with 400 volumes, 200 each. 
Many thanks~

**Reply by KumoLiu** [reply]
Hi @Harris91-lee,  I'd like to say that there are always many training tricks to try, but perhaps augmenting the data to increase the number of training sets is a good way to go. Of course, you can also experiment with different loss functions, optimizers, and so on. Thanks!

**Reply by Harris91-lee** [reply]
Hi @KumoLiu ^^
I also have one more question regarding the network used here for the classification.
Is Densnet121omptimum choice for the MR binary classification? like there are other networks resnet18, vgg16, or even cnn ...etc.  
Is there a reason behinde choosing Densnet121 over the other networks? 

Thank you in advance ^^

**Reply by KumoLiu** [reply]
Hi @Harris91-lee, I think there is probably no particular reason to use Densnet121 in the tutorial, just an example :). You could try to use any other networks as your baseline. And from the previous thread, I noticed that you have already had more than 95% accuracy. If your training can accept 100%, you could try to use vgg or any other simpler network. Reducing model complexity can sometimes improve model performance. Thanks.

**Reply by Harris91-lee** [reply]
I see ~ thank you for your response. I'm getting something between 96~98% now metric accuracy using Densenet121 which is great actually. However, im trying to see if i can get better results. I am willing to try just Densenet without '121' and as you mentioned vgg16. 

cheers ^^
